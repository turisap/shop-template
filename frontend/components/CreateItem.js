import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import ErrorMessage from '../components/ErrorMessage';
import { CONFIG } from '../config';



const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title : String!
        $description : String!
        $price : Int!
        $image : String
        $largeImage : String
    ) {
        createItem(
            title : $title
            description : $description
            price : $price
            image : $image
            largeImage : $largeImage
        ) {
            id
        }
    }
`;

class CreateItem extends Component {
    state = {
        title : '',
        description : '',
        image : '',
        largeImage : '',
        price : 0
    };

    handleChange = e => {
        const { name, type, value } = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({ [name] : val });
    };

    uploadFile = async e => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', CONFIG.CLOUDINARY_PRESET);

        const res = await fetch(CONFIG.CLOUDINARY_ENDPOINT, {
            method : "POST",
            body : data
        });

        const file = await res.json();
        this.setState({
            image : file.secure_url,
            largeImage : file.eager[0].secure_url
        })
    }

    render() {
        return (
            <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
                {(createItem, {loading, error}) => (
                    <Form
                        data-test="create-item-form"
                        onSubmit={async e => {
                        e.preventDefault();
                        // calling the mutation function and redirect user
                        const res = await createItem();
                        Router.push({
                            pathname : '/item',
                            query : { id : res.data.createItem.id }
                        })
                    }}>
                        <ErrorMessage error={error}/>
                        <fieldset disabled={loading} aria-busy={loading}>
                            <label htmlFor={'file'}>
                                Picture
                                <input
                                    type={'file'}
                                    id={'file'}
                                    name={'file'}
                                    placeholder={'Upload a picture'}
                                    required
                                    onChange={this.uploadFile}
                                />
                                {this.state.image && <img width="200" src={this.state.image} alt={"Upload preview"}/>}
                            </label>
                            <label htmlFor={'title'}>
                                Title
                                <input
                                    type={'text'}
                                    id={'title'}
                                    name={'title'}
                                    placeholder={'Title'}
                                    required
                                    value={this.state.title}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor={'price'}>
                                Price
                                <input
                                    type={'number'}
                                    id={'price'}
                                    name={'price'}
                                    placeholder={'Price'}
                                    required
                                    value={this.state.price}
                                    onChange={this.handleChange}
                                />
                            </label>
                            <label htmlFor={'price'}>
                                Description
                                <textarea
                                    id={'description'}
                                    name={'description'}
                                    placeholder={'Enter a description'}
                                    required
                                    value={this.state.description}
                                    onChange={this.handleChange}
                                />
                            </label>
                        </fieldset>
                        <button>SUBMIT</button>
                    </Form>
                )}
            </Mutation>
        );
    }
}

CreateItem.propTypes = {};

export default CreateItem;
export { CREATE_ITEM_MUTATION };
