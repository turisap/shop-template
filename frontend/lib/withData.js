import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { CONFIG } from '../config';

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? CONFIG.END_POINT : CONFIG.END_POINT,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
  });
}

export default withApollo(createClient);
