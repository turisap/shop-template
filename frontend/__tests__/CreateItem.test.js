import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from "../lib/testUtils";
import Router from 'next/router';

// mock the fetch API
const mockImage = 'http://lol.com/lol.jpg';
global.fetch = jest.fn().mockResolvedValue({
    json : () => ({
        secure_url : mockImage,
        eager : [{
            secure_url : mockImage
        }]
    })

});

describe('<CreateItem/>', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem/>
            </MockedProvider>
        );

        const form = wrapper.find('[data-test="create-item-form"]');
        //console.log(form.debug());
        expect(toJSON(form)).toMatchSnapshot();
    });

    it('uploads a file when changed', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem/>
            </MockedProvider>
        );

        const input = wrapper.find('input[type="file"]');
        //console.log(input.debug())
        input.simulate('change', { target : { files : ['fakedog.jpg'] }});
        await wait();
        const component = wrapper.find('CreateItem').instance();
        //console.log(component.state)
        expect(component.state.image).toEqual(mockImage);
        expect(component.state.largeImage).toEqual(mockImage);
        expect(global.fetch).toHaveBeenCalled();

        //removing the mock of fetch from the global object
        global.fetch.mockReset();
    });

    it('handles state updating', async () => {
        const wrapper = mount(
            <MockedProvider>
                <CreateItem/>
            </MockedProvider>
        );

        wrapper
            .find('#title')
            .simulate('change', { target : { value : 'Testing', name : 'title' }});
        wrapper
            .find('#price')
            .simulate('change', { target : { value : 534, name : 'price', type : 'number' }});
        wrapper
            .find('#description')
            .simulate('change', { target : { value : 'Piu piu lolipop', name : 'description' }});

        expect(wrapper.find('CreateItem').instance().state).toMatchObject({
            title : "Testing",
            price : 534,
            description : 'Piu piu lolipop'
        })
    })

    it('creates an item when te form is submitted', async() => {
        const item = fakeItem();
        const mocks = [
            {
                request : {
                    query : CREATE_ITEM_MUTATION,
                    variables : {
                        title : item.title,
                        description : item.description,
                        image : '',
                        largeImage : '',
                        price : item.price
                    }
                },
                result : {
                    data : {
                        createItem : {
                            ...item,
                            id : 'adclol',
                            __typename : 'Item'
                        }
                    }
                }
            }
        ];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <CreateItem/>
            </MockedProvider>
        );

        // simulate filling the form
        wrapper
            .find('#title')
            .simulate('change', { target : { value : item.title, name : 'title' }});
        wrapper
            .find('#price')
            .simulate('change', { target : { value : item.price, name : 'price', type : 'number' }});
        wrapper
            .find('#description')
            .simulate('change', { target : { value : item.description, name : 'description' }});

        Router.router = {
            push : jest.fn()
        };
        wrapper.find('form').simulate('submit');
        await wait(50);
        expect(Router.router.push).toHaveBeenCalled();
        expect(Router.router.push).toHaveBeenCalledWith({
                pathname: '/item',
                query : {
                    id: 'adclol'
                }
        });
    })
});