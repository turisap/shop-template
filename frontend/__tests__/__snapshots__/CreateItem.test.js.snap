// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> renders and matches snapshot 1`] = `
Array [
  <Form
    data-test="create-item-form"
    onSubmit={[Function]}
  >
    <form
      className="Form-sc-1ns280c-0 fTwjpq"
      data-test="create-item-form"
      onSubmit={[Function]}
    >
      <DisplayError
        error={Object {}}
      />
      <fieldset
        aria-busy={false}
        disabled={false}
      >
        <label
          htmlFor="file"
        >
          Picture
          <input
            id="file"
            name="file"
            onChange={[Function]}
            placeholder="Upload a picture"
            required={true}
            type="file"
          />
        </label>
        <label
          htmlFor="title"
        >
          Title
          <input
            id="title"
            name="title"
            onChange={[Function]}
            placeholder="Title"
            required={true}
            type="text"
            value=""
          />
        </label>
        <label
          htmlFor="price"
        >
          Price
          <input
            id="price"
            name="price"
            onChange={[Function]}
            placeholder="Price"
            required={true}
            type="number"
            value={0}
          />
        </label>
        <label
          htmlFor="price"
        >
          Description
          <textarea
            id="description"
            name="description"
            onChange={[Function]}
            placeholder="Enter a description"
            required={true}
            value=""
          />
        </label>
      </fieldset>
      <button>
        SUBMIT
      </button>
    </form>
  </Form>,
  <form
    className="Form-sc-1ns280c-0 fTwjpq"
    data-test="create-item-form"
    onSubmit={[Function]}
  >
    <DisplayError
      error={Object {}}
    />
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <label
        htmlFor="file"
      >
        Picture
        <input
          id="file"
          name="file"
          onChange={[Function]}
          placeholder="Upload a picture"
          required={true}
          type="file"
        />
      </label>
      <label
        htmlFor="title"
      >
        Title
        <input
          id="title"
          name="title"
          onChange={[Function]}
          placeholder="Title"
          required={true}
          type="text"
          value=""
        />
      </label>
      <label
        htmlFor="price"
      >
        Price
        <input
          id="price"
          name="price"
          onChange={[Function]}
          placeholder="Price"
          required={true}
          type="number"
          value={0}
        />
      </label>
      <label
        htmlFor="price"
      >
        Description
        <textarea
          id="description"
          name="description"
          onChange={[Function]}
          placeholder="Enter a description"
          required={true}
          value=""
        />
      </label>
    </fieldset>
    <button>
      SUBMIT
    </button>
  </form>,
]
`;
